integration-tests/tests/src/tests/sync/user.ts

```diff
@@ -0,0 +1,620 @@
+  expect(typeof user).equals("object");
+  expect(typeof user.accessToken).equals("string");
+  expect(typeof user.refreshToken).equals("string");
+  expect(typeof user.id).equals("string");
+  expect(typeof user.identities).equals("object");
+  expect(typeof user.customData).equals("object");
+  expect(user).instanceOf(Realm.User);
+}
+
+function assertIsSameUser(value: Realm.User, user: Realm.User | null) {
+  assertIsUser(value);
+  expect(value.accessToken).equals(user?.accessToken);
+  expect(value.id).equals(user?.id);
+}
+
+function assertIsError(error: any, message: string) {
+  expect(error).instanceOf(Error, "The API should return an Error");
+  if (message) {
+    expect(error.message).equals(message);
+  }
+}
+
+function assertIsAuthError(error: any, code: number, title: string) {
+  expect(error).instanceOf(Realm.App.Sync.AuthError, "The API should return an AuthError");
+  if (code) {
+    expect(error.code).equals(code);
+  }
+  if (title) {
+    expect(error.title).equals(title);
+  }
+}
+
+async function registerAndLogInEmailUser(app: Realm.App) {
+  const validEmail = randomVerifiableEmail();
+  const validPassword = "test1234567890";
+  await app.emailPasswordAuth.registerUser({ email: validEmail, password: validPassword });
+  const user = await app.logIn(Realm.Credentials.emailPassword(validEmail, validPassword));
+  assertIsUser(user);
+  assertIsSameUser(user, app.currentUser);
+  return user;
+}
+
+async function removeExistingUsers(app: Realm.App) {
+  const users = app.allUsers;
+  Object.keys(app.allUsers).forEach(async (id) => {
+    await app.removeUser(users[id]);
+  });
+}
+
+describe.skipIf(environment.missingServer, "User", () => {
+  describe("email password", () => {
+    importAppBefore("with-email-password");
+    it("login without username throws", async function (this: Mocha.Context & AppContext & RealmContext) {
+      // @ts-expect-error test logging in without providing username.
+      expect(() => Realm.Credentials.emailPassword(undefined, "password")).throws(
+        Error,
+        "email must be of type 'string', got (undefined)",
+      );
+    });
+
+    it("login without password throws", async function (this: Mocha.Context & AppContext & RealmContext) {
+      const username = uuid();
+      // @ts-expect-error test logging in without providing password.
+      expect(() => Realm.Credentials.emailPassword(username, undefined)).throws(
+        Error,
+        "password must be of type 'string', got (undefined)",
+      );
+    });
+
+    it("login with non existing user throws", async function (this: Mocha.Context & AppContext & RealmContext) {
+      const credentials = Realm.Credentials.emailPassword("foo", "pass");
+      return this.app
+        .logIn(credentials)
+        .then((user) => {
+          throw new Error("Login should have failed");
+        })
+        .catch((err) => {
+          expect(err.message).equals("invalid username/password");
+        });
```

Comment:
This (and a lot of the other tests) could fairly easily be refactored (now or later) to use "chai-as-promised":

```suggestion
      const credentials = Realm.Credentials.emailPassword("foo", "pass");
      await expect(this.app.logIn(credentials)).should.be.rejectedWith("invalid username/password");
```